### PR TITLE
Allow type T in WithFieldValue<T>

### LINKED ISSUE #5661: `T` isn't assignable to `WithFieldValue<T>`
### [REQUIRED] Describe your environment

  * Operating System version: -
  * Browser version: -
  * Firebase SDK version: 9.1.3
  * Firebase Product: Firestore
  * Typescript 4.4.4

#### Steps to reproduce:

Create a function which takes an argument of type `T`, and inside it call any firestore functions that receive `WithFieldValue` or `PartialWithFieldValue` (e.g. `setDoc`, `addDoc`, etc) with that value `T`.

Result is that TypeScript will say nope:

> Argument of type 'T' is not assignable to parameter of type 'WithFieldValue<T>'.

this creates a problem in that every call to these functions require a type cast, which is not ideal.

#### Relevant Code:

```typescript
class Bla<T> {
  withFieldValueT(value: WithFieldValue<T>) {
    
  }

  withT(value: T) {
    // doesn't work
    this.withFieldValueT(value);

    // works but has cast
    this.withFieldValueT(value as unknown as WithFieldValue<T>);
  }
}

const bla = new Bla<{ id: string, foo: number }>();
bla.withFieldValueT({ id: '', foo: 123 });
bla.withT({ id: '123', foo: 123 });
```

[TS playground link](https://www.typescriptlang.org/play?ts=4.4.4&ssl=17&ssc=1&pln=33&pc=36#code/PTAEFpK7oKBUAxAlgJwKYGcAuB7DoAxrgCbrxgxWSzYCeADuqAAqrIC2y2yAbswF5QOdgDsA5qAA+oAEa5cAG3QBDUdNCiArh1npUG7YsUatosgDNko9CQDcsWNez6LKwsxTpFJAGorFLWYAb1hQUGRMAFEARy0AgAoASgAuOQVlNQcAX0d6JlAAdW4ACy8ff0D0AB4AFQA+UCFa0HQADxdzTFZ2Lh5+UAB+UBa0lvbOkm7g7KHQUPDwgG0AaQj1AGt0OlwLEYBdNOLsMuRvPwCgutX9xplyi6qc0DSWFVQeALr6hwoIaigfwAwlocLgOERSOQEADAYRFCpMN0AEII77zMKgADupQelSCtQSvEu6COuLOFRJ3ySGMWoExuUxOJOhOJVTGNIWiwQJFwWFEAHJsNj8BtMeETpEAHTM07nfHoVkkpJ2UAIarURx0hBY0XdWRaYUlRFERHYcWgSWYGXk+UkpVVUAmswbUS4LHqE3HOWUqrUhzhXKM4iiHByBFNTToLGgVEqarBCIkNIiaziAA0oAsCjS2l0+lA2XqyQcsgRNpOePtCUTyGToAFAsz2dwaQAjAAmADMhZVsDLKgrJUJtfrAs7XabWZzoAnvbsQA)

### PR DESCRIPTION
**Pull request by thebrianchen**
Fixes #5661. Thanks @gustavohenke for proactively bringing up the fix!

This change allows developers to pass an object of type `T` into `WithFieldValue<T>`.

### CODE REVIEW

File: packages/firestore/test/lite/integration.test.ts [hunk 1]
```diff
@@ -1496,6 +1496,51 @@ describe('withConverter() support', () => {
           });
         });
       });
+
+      describe('used as a type', () => {
+        class ObjectWrapper<T> {
+          withFieldValueT(value: WithFieldValue<T>): void {
+            // eslint-disable-next-line no-console
+            console.log(value);
```
[the remark this answers is not in the document]
Comment (thebrianchen) [reply packages/firestore/test/lite/integration.test.ts]: Removed.

File: packages/firestore/src/lite-api/reference.ts [hunk 2]
```diff
@@ -54,22 +54,19 @@ export interface DocumentData {
...
-  ? { [K in keyof T]: WithFieldValue<T[K]> | FieldValue }
-  : Partial<T>;
-
+export type WithFieldValue<T> =
+  | T
+  | (T extends Primitive
+      ? T
+      : T extends {}
+      ? { [K in keyof T]: WithFieldValue<T[K]> | FieldValue }
+      : Partial<T>);
```
[the remark this answers is not in the document]
Comment (thebrianchen) [reply packages/firestore/src/lite-api/reference.ts]: This should technically be `never`. Fixed.

**Review by schmidt-sebastian — APPROVED** [review]

File: .changeset/short-mice-itch.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@firebase/firestore': minor
```
Comment (Feiyang1) [inline .changeset/short-mice-itch.md:2]:
```suggestion
'@firebase/firestore': minor
'firebase': minor
```

**Review by Feiyang1 — CHANGES_REQUESTED** [review]

File: .changeset/short-mice-itch.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+---
+
+Expanded `Firestore.WithFieldValue<T>` to include `T`. This allows developers to delegate `WithFieldValue<T>` inside a wrappers of type `T` to avoid exposing Firebase types beyond Firebase-specific logic.
```
Comment (egilmorez) [inline .changeset/short-mice-itch.md:5]: I think "a wrappers" should just be "wrappers"

**Review by egilmorez — APPROVED** [review]
Found one typo, otherwise LGTM!

(Reviewing for Mark, who is out today)

File: .changeset/short-mice-itch.md (hunk 4 above)
Comment (thebrianchen) [reply .changeset/short-mice-itch.md]: done. thanks for catching

**Review by Feiyang1 — APPROVED** [review]

**Pull request merged**